packages/core/factories/src/useUserFactory.ts

```diff
@@ -0,0 +1,113 @@
+import { ref, Ref, watch, computed } from '@vue/composition-api';
+import { UseUser } from '@vue-storefront/interfaces';
+
+export type UseUserFactoryParams<USER, CART, UPDATE_USER_PARAMS> = {
+  cart: Ref<CART>;
+  loadUser: () => Promise<USER>;
+  logOut: (params?: {currentUser?: USER}) => Promise<void>;
+  updateUser: (params: {currentUser: USER; paramsToUpdate: UPDATE_USER_PARAMS}) => Promise<USER>;
+  register: (params: { email: string; password: string; firstName?: string; lastName?: string }) => Promise<USER>;
+  logIn: (params: { username: string; password: string }) => Promise<({updatedUser: USER; updatedCart: CART})>;
+  changePassword: (params: {currentUser: USER; currentPassword: string; newPassword: string}) => Promise<USER>;
+};
+
+export function useUserFactory<USER, CART, UPDATE_USER_PARAMS>(
+  factoryParams: UseUserFactoryParams<USER, CART, UPDATE_USER_PARAMS>
+) {
+  const user: Ref<USER> = ref({});
+  const loading: Ref<boolean> = ref(false);
+  const isAuthenticated = computed(
+    () => user.value && Object.keys(user.value).length > 0
+  );
+
+  return function useUser(): UseUser<USER, UPDATE_USER_PARAMS> {
+    watch(user, async () => {
+      if (isAuthenticated.value) {
+        return;
+      }
+      loading.value = true;
+      try {
+        user.value = await factoryParams.loadUser();
+        loading.value = false;
+      } catch (err) {
+        console.log('useUser:Factory:Watch', err);
+        loading.value = false;
+      }
+    });
+
+    const updateUser = async (params: UPDATE_USER_PARAMS) => {
+      loading.value = true;
+      try {
+        user.value = await factoryParams.updateUser({currentUser: user.value, paramsToUpdate: params});
+        loading.value = false;
+      } catch (err) {
+        console.error('useUser:Factory:updateUser', err);
+        loading.value = false;
```

Comment:
loading=false should go to `finally` block
also, catch should rethrow error, because otherwise, the user doesn't know that something went wrong and cant react on that

Reply:
Done